File: src/cli/plan.ts
export const MAINSTREAM_TLDS = ["com", "net", "org", "io", "ai", "co", "app", "dev", "sh"] as const;

export type QueryGroup = {
  input: string;
  root: string;
  domains: string[];
  suggested: boolean;
};

const bareLabelPattern = /^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$/;

const hasTld = (value: string): boolean => value.includes(".");

const bareLabel = (value: string): boolean => bareLabelPattern.test(value);

export const createQueryPlan = (
  inputs: string[],
): { groups: QueryGroup[]; lookupDomains: string[] } => {
  const groups = inputs.map((input) => {
    if (!hasTld(input) && bareLabel(input)) {
      return {
        input,
        root: input,
        domains: MAINSTREAM_TLDS.map((tld) => `${input}.${tld}`),
        suggested: true,
      } satisfies QueryGroup;
    }

    return {
      input,
      root: input,
      domains: [input],
      suggested: false,
    } satisfies QueryGroup;
  });

  const lookupDomains = [...new Set(groups.flatMap((group) => group.domains))];

  return { groups, lookupDomains };
};

File: internal/cli/plan.go
package cli

import "regexp"

var mainstreamTLDs = []string{"com", "net", "org", "io", "ai", "co", "app", "dev", "sh", "so"}

type queryGroup struct {
	Input     string
	Root      string
	Domains   []string
	Suggested bool
}

var bareLabelRe = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?$`)

func hasTLD(value string) bool {
	for _, c := range value {
		if c == '.' {
			return true
		}
	}

	return false
}

func createQueryPlan(inputs []string) (groups []queryGroup, lookupDomains []string) {
	for _, input := range inputs {
		if !hasTLD(input) && bareLabelRe.MatchString(input) {
			domains := make([]string, len(mainstreamTLDs))
			for i, tld := range mainstreamTLDs {
				domains[i] = input + "." + tld
			}
			groups = append(groups, queryGroup{
				Input:     input,
				Root:      input,
				Domains:   domains,
				Suggested: true,
			})
		} else {
			groups = append(groups, queryGroup{
				Input:     input,
				Root:      input,
				Domains:   []string{input},
				Suggested: false,
			})
		}
	}

	seen := make(map[string]bool)
	for _, g := range groups {
		for _, d := range g.Domains {
			if !seen[d] {
				seen[d] = true
				lookupDomains = append(lookupDomains, d)
			}
		}
	}

	return groups, lookupDomains
}
